test/helpers/value_hook.dart

```diff
@@ -1,42 +1,198 @@
 import 'dart:async';
 
-import 'package:test/test.dart' as test show expect, Matcher;
-
-import 'wait_for.dart';
+import 'package:test/test.dart' as test show expect, Matcher, TestFailure;
 
 /// Useful to easily test handler functions
 /// that set values outside of their scope.
 ///
 /// ```dart
-/// // Construct VHook with initial value of 'false'
-/// VHook<bool> calledHandler = VHook<bool>(false);
+/// // Construct VHook with no initial value
+/// VHook calledHandler = VHook();
 ///
 /// willCallHandler(() {
-///   calledHandler.set(true);
 ///   // Some handler code
+///   // ...
+///   calledHandler.complete();
 /// });
 ///
 /// // Test that the handler was called
-/// // This will throw an error if the handler was not called.
-/// calledHandler.expect(equals(true));
+/// // If the handler was not called after 5s, throw an error
+/// calledHandler.awaitCompletion(Duration(seconds: 5));
 /// ```
 class VHook<T> {
   /// The value of the variable.
-  T _value;
+  dynamic _value;
```

Comment:
I'm not a big fan of using `dynamic` for this. I'd recommend using `late T _value` and an additional `_hasValue` flag. You could add a getter and setter to throw a custom error when accessing `_value` when `_hasValue` is false. Maybe even a small utility class

Reply:
We _have_ to use dynamic here. The default is `_VHookValues.none` which internally represents a `VHook` without any value (this allows the value `none`). So this can assume any value of type `T` or `VHookValues.none` which means it has to be dynamic. The getter does check the value though and will always return `T` which makes this an internal dynamic only.